tot/data_processing/scaler.py

```diff
@@ -0,0 +1,219 @@
+        """
+        if fit:
+            self.transformer.fit(df["y"].values.reshape(-1, 1))
+        df["y"] = self.transformer.transform(df["y"].values.reshape(-1, 1))
+        return df
+
+    def _rescale_per_series(self, df: pd.DataFrame, col_name: str) -> pd.DataFrame:
+        """
+        Applies `inverse_transform` per series. First, pivot is performed on the dataframe so that unique `ID`s
+        become columns, then inverse transformation is applied. Data is returned in its original format.
+
+        Parameters:
+        -----------
+        df : pd.DataFrame
+            dataframe containing column ``ds``, ``y``, [``yhat<i>``], and optionally ``ID`` with data
+        col_name : str
+            name of the column, on which the operation is applied
+
+        Returns:
+        --------
+        pd.DataFrame
+            dataframe containing column ``ds``, ``y``, [``yhat<i>``], and optionally ``ID`` with rescaled data
+        """
+        IDs = df["ID"].unique()
+        df_pivot = _pivot(df, col_name)
+
+        df_pivot[IDs] = self.transformer.inverse_transform(df_pivot[IDs])
+
+        return _melt(df_pivot, IDs, col_name)
+
+    def _rescale(self, df: pd.DataFrame, col_name: str) -> pd.DataFrame:
+        """
+        Applies `inverse_transform` on column `col_name` in `df`.
+
+        Parameters:
+        -----------
+        df : pd.DataFrame
+            dataframe containing column ``ds``, ``y``, [``yhat<i>``], and optionally ``ID`` with data
+        col_name : str
+            name of the column, on which the operation is applied
+
+        Returns:
+        --------
+        pd.DataFrame
+            dataframe containing column ``ds``, ``y``, [``yhat<i>``],  and optionally ``ID`` with rescaled data
+        """
+        df[col_name] = self.transformer.inverse_transform(df[col_name].values.reshape(-1, 1))
+        return df
+
+    def _inverse_transform(self, df: pd.DataFrame) -> pd.DataFrame:
+        """
+        Applies rescaling on the `df`. First, rescaling is performed on the `y` column to create the main df. Then,
+        operation is repeated on all `yhat` values and results are updated in the main df. Proper `rescale`
+        implementation is chosen based on scaling level.
+
+        Parameters:
+        -----------
+        df : pd.DataFrame
+            dataframe containing column ``ds``, ``y``, [``yhat<i>``] and optionally ``ID`` with data
+
+        Returns:
+        --------
+        pd.DataFrame
+            dataframe containing column ``ds``, ``y``, [``yhat<i>``] and optionally ``ID`` with rescaled data
+        """
+        if self.scaling_level == "per_time_series":
+            rescale_method = self._rescale_per_series
+        else:
+            rescale_method = self._rescale
+
+        result = rescale_method(df, "y")
+
+        yhats = [col for col in df.columns if "yhat" in col]
+        for yhat in yhats:
+            result[yhat] = rescale_method(df, yhat)[yhat]
```

Comment:
Can we avoid passing the column names?

Reply:
`inverse_transform` is done on a 2d array (timestamps x IDs) and `y`s (and `yhat<i>`s) are values in this array. I'm not sure how to perform rescaling on all of the values in one step, because it would either result in having a 2d array with list of values or a <i+1>d array, neither of them can be consumed by `inverse_transform`. Lmk if I missunderstood something